crates/ra_syntax/src/grammar/expressions.rs

```diff
@@ -158,18 +158,18 @@ fn current_op(p: &Parser) -> (u8, Op) {
 // Parses expression with binding power of at least bp.
 fn expr_bp(p: &mut Parser, r: Restrictions, bp: u8) -> BlockLike {
     let mut lhs = match lhs(p, r) {
-        Some(lhs) => {
+        (Some(lhs), blocklike) => {
             // test stmt_bin_expr_ambiguity
             // fn foo() {
             //     let _ = {1} & 2;
             //     {1} &2;
             // }
-            if r.prefer_stmt && is_block(lhs.kind()) {
+            if r.prefer_stmt && (is_block(lhs.kind()) || blocklike == Some(BlockLike::Block)) {
```

Comment:
`is_block(lhs.kind())` check should be redundant, given that we return `blockfile` from lhs() call